src/Affjax/Driver.purs

```diff
@@ -0,0 +1,231 @@
+    , password: toNullable req.password
+    , withCredentials: req.withCredentials
+    , timeout: fromMaybe 0.0 $ (\(Milliseconds x) -> x) <$> req.timeout
+    }
+
+  extractContent :: RequestBody.RequestBody -> Either String Foreign
+  extractContent = case _ of
+    RequestBody.ArrayView f ->
+      Right $ f (unsafeToForeign :: forall b. ArrayView b -> Foreign)
+    RequestBody.Blob x ->
+      Right $ (unsafeToForeign :: Blob -> Foreign) x
+    RequestBody.Document x ->
+      Right $ (unsafeToForeign :: Document -> Foreign) x
+    RequestBody.String x ->
+      Right $ (unsafeToForeign :: String -> Foreign) x
+    RequestBody.FormData x ->
+      Right $ (unsafeToForeign :: FormData -> Foreign) x
+    RequestBody.FormURLEncoded x -> do
+      note "Body contains values that cannot be encoded as application/x-www-form-urlencoded"
+        $ (unsafeToForeign :: String -> Foreign) <$> FormURLEncoded.encode x
+    RequestBody.Json x ->
+      Right $ (unsafeToForeign :: String -> Foreign) (J.stringify x)
+
+  headers :: Maybe RequestBody.RequestBody -> Array RequestHeader
+  headers reqContent =
+    addHeader (ContentType <$> (RequestBody.toMediaType =<< reqContent))
+      $ addHeader (Accept <$> ResponseFormat.toMediaType req.responseFormat) req.headers
+
+  timeoutErrorMessageIdent :: String
+  timeoutErrorMessageIdent = "AffjaxTimeoutErrorMessageIdent"
+
+  requestFailedMessageIdent :: String
+  requestFailedMessageIdent = "AffjaxRequestFailedMessageIdent"
+
+  addHeader :: Maybe RequestHeader -> Array RequestHeader -> Array RequestHeader
+  addHeader mh hs = case mh of
+    Just h | not $ any (on eq RequestHeader.name h) hs -> hs `Arr.snoc` h
+    _ -> hs
+
+  parseJSON :: String -> F Json
+  parseJSON = case _ of
+    "" -> pure J.jsonEmptyObject
+    str -> either (fail <<< ForeignError) pure (jsonParser str)
+
+  fromResponse :: Foreign -> F a
+  fromResponse = case req.responseFormat of
+    ResponseFormat.ArrayBuffer _ -> unsafeReadTagged "ArrayBuffer"
+    ResponseFormat.Blob _ -> unsafeReadTagged "Blob"
+    ResponseFormat.Document _ -> \x ->
+      unsafeReadTagged "Document" x
+        <|> unsafeReadTagged "XMLDocument" x
+        <|> unsafeReadTagged "HTMLDocument" x
+    ResponseFormat.Json coe -> coe <<< parseJSON <=< unsafeReadTagged "String"
+    ResponseFormat.String _ -> unsafeReadTagged "String"
+    ResponseFormat.Ignore coe -> const $ coe (pure unit)
+
+type AjaxRequest :: Type -> Type
+type AjaxRequest a =
+  { method :: String
+  , url :: URL
+  , headers :: Array { field :: String, value :: String }
+  , content :: Nullable Foreign
+  , responseType :: String
+  , username :: Nullable String
+  , password :: Nullable String
+  , withCredentials :: Boolean
+  , timeout :: Number
+  }
+
+foreign import data Xhr :: Type
+
+-- Drivers should have the following 'shape':
+-- ```
+-- { newXHR :: Effect Xhr
+-- , fixupUrl :: Fn2 Xhr String String
+-- }
+-- ```
+foreign import data AffjaxDriver :: Type
```

Comment:
I don't think we need this `Xhr` type, since it's not actually used anywhere aside from within this comment?